botcore/perms.py

```diff
@@ -1,83 +1,62 @@
-    pass
-
-class NotGuildMember(CheckFailure):
-    pass
-
-class NotAllowedChannel(CheckFailure):
-    pass
-
-class NotAdminChannel(CheckFailure):
-    pass
-
-class NotDMChannel(CheckFailure):
-    pass
-
-def check(predicate):
+def check(predicate, error):
     """
     Decorator. Method is only handled if it passes a check in the form of the
     function predicate.
     Can only be used with cogs.
 
     Args:
-        predicate: A function that takes in the cog and the context as arguments
-            and returns a boolean value representing the result of a check.
+        predicate: A function that takes in the cog and the context as 
+            arguments and returns a boolean value representing the result of a
+            check. Does not need to be async, but can be.
+        error: Whether to send error message if check fails.
     """
 
+    async_predicate = predicate
+    if not iscoroutinefunction(predicate):
+        @wraps(predicate)
+        async def wrapper(cog, ctx):
+            return predicate(cog, ctx)
+        async_predicate = wrapper
+
     def decorator(func):
-        @functools.wraps(func)
+        @wraps(func)
         async def wrapper(*args, **kwargs):
             cog = args[0]
             ctx = args[1]
-            if not predicate(cog, ctx):
+            
+            res, err_msg = await async_predicate(cog, ctx)
+            if not res:
+                if error:
+                    await ctx.send(err_msg)
                 return
             return await func(*args, **kwargs)
         return wrapper
 
-    if inspect.iscoroutinefunction(predicate):
-        decorator.predicate = predicate
-    else:
-        @functools.wraps(predicate)
-        async def wrapper(ctx):
-            return predicate(ctx)
-        decorator.predicate = wrapper
-
     return decorator
 
 def is_verified_user(error=False):
     """
-    Decorator. Only allows method to execute if invoker has the verified role as
-    defined in the config.
+    Decorator. Only allows method to execute if invoker has the verified role 
+    as defined in the config.
     Cog must have bot as an instance variable.
 
     Args:
-        error: Whether to raise error if check fails.
-
-    Raises:
-        NotVerified: Check failed and error == True.
+        error: Whether to send error message if check fails.
     """
 
     def predicate(cog, ctx):
-        member = cog.bot.get_guild(config["server-id"]).get_member(ctx.author.id)
-        if member is None or config["verified-role"] not in map(lambda r: r.id, member.roles):
-            if error:
-                raise NotVerified("You must be verified to do that.")
-            return False
-        return True
-    return check(predicate)
+        member = _get_member(cog, ctx.author.id)
+        if member is None or config["verified-role"] \
+            not in _get_role_ids(member):
+            return False, "You must be verified to do that."
+        return True, None
```

Comment:
not a fan of returning a tuple for the sake of being able to return multiple values. consider a custom class. (PCSoc discord bot handles this well)